substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -0,0 +1,1591 @@
+	/// view.
+	///
+	/// If there are no views, and mempool transaction is reported as invalid for the given view,
+	/// the transaction is reported as invalid and removed from the mempool. This does not apply to
+	/// stale and temporarily banned transactions.
+	///
+	/// As the listeners for watched transactions were registered at the very beginning of maintain
+	/// procedure (`register_listeners`), this function accepts the list of watched transactions
+	/// from the mempool for which listener was actually registered to avoid submit/maintain races.
+	async fn update_view_with_mempool(
+		&self,
+		view: &View<ChainApi>,
+		watched_xts: Vec<(ExtrinsicHash<ChainApi>, Arc<TxInMemPool<ChainApi, Block>>)>,
+	) {
+		log::debug!(
+			target: LOG_TARGET,
+			"update_view_with_mempool: {:?} xts:{:?} v:{}",
+			view.at,
+			self.mempool.unwatched_and_watched_count(),
+			self.active_views_count()
+		);
+		//todo: this could be collected/cached in view
+		let included_xts = self.extrinsics_included_since_finalized(view.at.hash).await;
+
+		//todo: can we do better - w/o clone?
+		let xts = self.mempool.clone_unwatched();
+
+		let mut all_submitted_count = 0;
+		if !xts.is_empty() {
+			let unwatched_count = xts.len();
+			let mut buckets = HashMap::<TransactionSource, Vec<ExtrinsicFor<ChainApi>>>::default();
+			xts.into_iter()
+				.filter(|(hash, _)| !view.pool.validated_pool().pool.read().is_imported(hash))
+				.filter(|(hash, _)| !included_xts.contains(&hash))
+				.map(|(_, tx)| (tx.source(), tx.tx()))
+				.for_each(|(source, tx)| buckets.entry(source).or_default().push(tx));
+
+			for (source, xts) in buckets {
+				all_submitted_count += xts.len();
+				let _ = view.submit_many(source, xts).await;
+			}
+			log::debug!(target: LOG_TARGET, "update_view_with_mempool: at {:?} unwatched {}/{}", view.at.hash, all_submitted_count, unwatched_count);
+		}
+
+		let watched_submitted_count = watched_xts.len();
+
+		let mut buckets = HashMap::<
+			TransactionSource,
+			Vec<(ExtrinsicHash<ChainApi>, ExtrinsicFor<ChainApi>)>,
+		>::default();
+		watched_xts
+			.into_iter()
+			.map(|(tx_hash, tx)| (tx.source(), tx_hash, tx.tx()))
+			.for_each(|(source, tx_hash, tx)| {
+				buckets.entry(source).or_default().push((tx_hash, tx))
+			});
+
+		let mut watched_results = Vec::default();
+		for (source, watched_xts) in buckets {
+			let hashes = watched_xts.iter().map(|i| i.0).collect::<Vec<_>>();
+			let results = view
+			.submit_many(source, watched_xts.into_iter().map(|i| i.1))
+			.await
+			.into_iter()
+			.zip(hashes)
+			.map(|(result, tx_hash)| {
+				let result = result.or_else(
+					|error| {
+						let error = error.into_pool_error();
+						log::trace!(
+							target: LOG_TARGET,
+							"[{:?}] update_view_with_mempool: submit (watched) result: {:?} {:?}",
+							tx_hash,
+							view.at.hash,
+							error,
+						);
+						match error {
+							Ok(
+								Error::InvalidTransaction(InvalidTransaction::Stale) | Error::TemporarilyBanned
```

Comment:
Some comment on why these are seen as acceptable would be nice.

Reply:
It was leftover from: https://github.com/paritytech/polkadot-sdk/blob/c28c1eec92072849ebb492141163f619d2d04d5c/substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs#L1012

Removed in 872cd88fac7